libsignal-protocol/src/store/identity_key_store.rs

```diff
@@ -40,20 +55,38 @@ pub(crate) fn new_vtable<I: IdentityKeyStore + 'static>(
     }
 }
 
-struct State(Box<dyn IdentityKeyStore>);
+struct State(Box<dyn IdentityKeyStore<'static>>);
 
 unsafe extern "C" fn get_identity_key_pair(
-    _public_data: *mut *mut sys::signal_buffer,
-    _private_data: *mut *mut sys::signal_buffer,
-    _user_data: *mut c_void,
+    public_data: *mut *mut sys::signal_buffer,
+    private_data: *mut *mut sys::signal_buffer,
+    user_data: *mut c_void,
 ) -> c_int {
-    unimplemented!()
+    assert!(!user_data.is_null());
+    assert!(!public_data.is_null());
+    assert!(!private_data.is_null());
+
+    let user_data = &*(user_data as *const State);
+
+    match user_data.0.identity_key_pair() {
+        // A funny way to pattern match :)
```

Comment:
Yeah, the nested `match` statements look kinda funny. I feel like we could do with a `try!()` macro... What are your thoughts on something like this?

```rust
macro_rules try_e {
  ($thing:expr) => {
    match $thing {
      Ok(got) => got,
      Err(e) => return e.code(),
    }
  }
}
```

Then we'd be able to rewrite it as

```rust
let id = try_e!(user_data.0.identity_key_pair());
let public_key = try_e!(id.public().serialize());
let private_key = try_e!(id.private().serialize());

*public_data = public_key.into_raw();
*private_data = private_key.into_raw();

sys::SG_SUCCESS as c_int
```

And if you can come up with a better name, feel free to change it!

Reply:
The problem is that the `serialize()` method does not have the `InternalError` type, as i remember i think it has `failure::Error`.